globeandmail/sophi-for-wordpress
# Enhancement/328 Sophi Overrides API Integration

### Description of the Change

* Integrated the [Sophi Override functions](https://documenter.getpostman.com/view/16875817/TzzHmCyw#2a201cca-c05b-4d98-b9fd-5e96ce1baf54) (add, remove, replace, ban) into the existing Site Automation block.
* The existing block will not break after this integration because the final code snippet of the block does not change! (i.e. nothing to worry about the migration of existing block).


Closes #328 & Closes #216

### Alternate Designs

_List view:_

![image](https://user-images.githubusercontent.com/25176325/195886455-dd835f3e-7f14-41dd-884b-1061a3ed0d80.png)
[image link](https://www.screencast.com/t/zEMLZpusYI9)

---

_The 4 Override buttons:_

![image](https://user-images.githubusercontent.com/25176325/196750851-b8085ade-cc65-4199-bd17-ec8ddff97c48.png)
[image link](https://www.screencast.com/t/0r60bXIIZo)

---

_Add/Replace Override Popup:_

![image](https://user-images.githubusercontent.com/25176325/196751550-79ed0550-a5e2-4d66-b3d3-bb64a85f6a21.png)
[image link](https://www.screencast.com/t/7cSwkE1pC)

---

_Out/Ban Override Popup:_

![image](https://user-images.githubusercontent.com/25176325/196752970-6ef17c15-77b6-4f33-bae9-789e5a4063e6.png)
[image link](https://www.screencast.com/t/crkbQS4H)

---
_Settings page:_

![image](https://user-images.githubusercontent.com/25176325/196751950-2e73fa7c-26fb-41f6-ad14-a5539f5c41c7.png)
[image link](https://www.screencast.com/t/JsZlqyDcG7K)

---

_When posts are received from the Sophi server, but none of them exist in the WordPress site:_

![image](https://user-images.githubusercontent.com/25176325/197990551-015dcc85-14f6-435a-87b4-621c20647da4.png)
[image link](https://www.screencast.com/t/Np9gAvUAtGB)

---

_When posts are received from the Sophi server, but none of them exist in the WordPress site:_

![image](https://user-images.githubusercontent.com/25176325/197990723-84c51db7-c2dd-41f4-ba68-12c9c0712a65.png)
[image link](https://www.screencast.com/t/I4nYVhKM6s)

### Possible Drawbacks


### Verification Process

1. Make sure you have configured proper credential details on the Sophi settings page.
2. Add a Site Automation block to a page.
3. Fill in "Page name" and "Widget name" in the block settings on the right sidebar.
4. See the curated post list, it should appear.
5. Select any individual post item and see it has its own settings (`Blockcontrol`) at the top.
6. Click the gear icon, and a popup should appear.
7. Perform Add/Replace/Remove/Ban actions by filling out the form.
8. For add/replace, the result will appear quickly.
9. But for remove/ban, the Sophi server will take some time to reflect.


### Checklist:





### Changelog Entry

> Added - Override feature to the Sophi Site Automation Block.
> Added - `???` hook to allow for code-based Override feature integrations.
> Changed - Sophi Site Automation Block from a server-side-rendered custom block to a nested block.

### Credits

Props @jeffpaul @fabiankaegy @faisal-alvi
